lib/db.c

```diff
@@ -74,12 +74,54 @@ static const uint32_t GRN_TABLE_PAT_KEY_CACHE_SIZE = 1 << 15;
       grn_string_get_normalized(ctx, nstr, &key, &key_size, NULL);\
       block\
       grn_obj_close(ctx, nstr);\
+    } else {\
+      char name[GRN_TABLE_MAX_KEY_SIZE];\
+      int name_size;\
+      name_size = grn_obj_name(ctx, (grn_obj *)(table), name, GRN_TABLE_MAX_KEY_SIZE);\
+      ERR(GRN_INVALID_ARGUMENT,\
+          "[key][normalize] failed to normalize: <%.*s>: <%.*s>",\
+          name_size, name,\
+          (int)key_size, (const char *)key);\
     }\
   } else {\
     block\
   }\
 } while (0)
 
+static grn_obj *
+grn_inspect_key_with_table_name(grn_ctx *ctx, grn_obj *buffer, grn_obj *table, const void *key, unsigned int key_size)
+{
+  grn_obj inspected;
+
+  GRN_TEXT_INIT(&inspected, 0);
+  grn_inspect_limited(ctx, &inspected, table);
+
+  if (!table) {
+    grn_text_printf(ctx, buffer, "table must not NULL: <%.*s>",
+                 (int)GRN_TEXT_LEN(&inspected), GRN_TEXT_VALUE(&inspected));
+    GRN_OBJ_FIN(ctx, &inspected);
+    return buffer;
+  }
+
+  if (key && key_size > 0) {
+    grn_obj key_buffer;
+    GRN_OBJ_INIT(&key_buffer, GRN_BULK, GRN_OBJ_DO_SHALLOW_COPY, table->header.domain);
+    GRN_TEXT_SET(ctx, &key_buffer, key, key_size);
+    grn_inspect(ctx, buffer, &key_buffer);
+    GRN_OBJ_FIN(ctx, &key_buffer);
+    grn_text_printf(ctx, buffer, "<%.*s>: <%.*s>",
+                    (int)GRN_TEXT_LEN(&inspected), GRN_TEXT_VALUE(&inspected),
+                    (int)GRN_TEXT_LEN(&key_buffer), GRN_TEXT_VALUE(&key_buffer));
```

Comment:
Why should  we need to append `key_buffer` content to `buffer` twice?

Reply:
Oh, It is wrong...